crates/matrix-sdk-ui/src/timeline/queue.rs

```diff
@@ -96,78 +101,104 @@ pub(super) async fn send_queued_messages(
     info!("Stopped");
 }
 
-async fn handle_message(
+async fn send_or_queue_msg(
     msg: LocalMessage,
     room: Room,
     send_task: &mut SendMessageTask,
     queue: &mut VecDeque<LocalMessage>,
-    timeline_inner: &TimelineInner,
+    timeline: &TimelineInner,
 ) {
-    if queue.is_empty() && send_task.is_idle() {
-        if room.state() == RoomState::Joined {
-            send_task.start(room, timeline_inner.clone(), msg);
-        } else {
-            info!("Refusing to send message, room is not joined");
-            timeline_inner
-                .update_event_send_state(
-                    &msg.txn_id,
-                    EventSendState::SendingFailed {
-                        // FIXME: Probably not exactly right
-                        error: Arc::new(matrix_sdk::Error::InconsistentState),
-                    },
-                )
-                .await;
-        }
-    } else {
+    // Only send the message immediately if there aren't other messages to be sent
+    // first, and we're not currently sending a message.
+    if !queue.is_empty() || !send_task.is_idle() {
         queue.push_back(msg);
+        return;
+    }
+
+    if room.state() != RoomState::Joined {
+        info!("Refusing to send message, room is not joined");
+        timeline
+            .update_event_send_state(
+                &msg.txn_id,
+                EventSendState::SendingFailed {
+                    // FIXME: Probably not exactly right
+                    error: Arc::new(matrix_sdk::Error::InconsistentState),
+                },
+            )
+            .await;
+        return;
```

Comment:
Why would you do a `if/return/implicit return` instead of a `if/else` here? I understand the above one to avoid a nested if, this one seems to go a bit too far with the removal of the nesting.

Reply:
I do like that the main operation is not under a conditional, and just lives at the end in the main path, but I can revert this one change here.